hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestBlockOutputStream.java

```diff
@@ -305,24 +305,24 @@ void testWriteExactlyFlushSize(boolean flushDelay) throws Exception {
       assertEquals(1, keyOutputStream.getStreamEntries().size());
       // The previously written data is equal to flushSize, so no action is
       // triggered when execute flush, if flushDelay is enabled.
-      assertEquals(pendingWriteChunkCount + (flushDelay ? 2 : 0),
-          metrics.getPendingContainerOpCountMetrics(WriteChunk));
-      assertEquals(pendingPutBlockCount + (flushDelay ? 1 : 0),
-          metrics.getPendingContainerOpCountMetrics(PutBlock));
+      if (!flushDelay) {
```

Comment:
Thanks @chungen0126 for the patch.

Would you mind putting a paragraph of brief explanation in the PR description on how this test was broken, and especially why the following checks should be skipped when `flushDelay == true`?

Reply:
Why was this conversation resolved?  I don't see the explanation in the PR description.

Reply:
After check, `assertEquals(0, blockOutputStream.getBufferPool().computeBufferData());
        assertEquals(dataLength, blockOutputStream.getTotalAckDataLength());` these two assertions don't need to be changed.
And I add some description explaining why those pendingContainerOp assersions should be changed

Reply:
Sorry, I missed the comment added in the code, was looking for it in the PR (or Jira) description.

However,

> If flushDelay is disabled, it will call waitOnFlushFutures to wait all putBlocks finished.

I don't think that explains why the test started failing suddenly around Mar 15, 2024 (see list of failures in HDDS-10557).  Last non-trivial change was 493c90066f on Dec 15, 2023.

Reply:
@chungen0126 Let me clarify my comment.  The test has been passing without intermittent failures for a long time.  Some change in either Ozone production code or the test environment must have triggered the failures.  We need to find this change.  I could be wrong, but I suspect it was 2f05353940 (committed Mar 6), which introduced multi-threaded execution in `BlockOutputStream` response processing.

Reply:
Tried 10x30 iterations:
- all [passed](https://github.com/adoroszlai/ozone/actions/runs/8845777511) with commit preceding HDDS-10384
- 2 [failures](https://github.com/adoroszlai/ozone/actions/runs/8845790712/job/24293087707) with HDDS-10384

Reply:
I see, I'll check the commit again.

Reply:
@adoroszlai I think the failures are not caused by [HDDS-10384](https://issues.apache.org/jira/browse/HDDS-10384). I run 10x30 iterations without [HDDS-10384](https://issues.apache.org/jira/browse/HDDS-10384), and it failed in my [test](https://github.com/chungen0126/ozone/actions/runs/8974559632).
I think the failures is due to that the `WriteChunk` and `PutBlock` completed before the assertion. One way to reproduce the failure is to add a some sleep after `write()`. 
In `testWriteExactlyFlushSize`, there is no reason that `WriteChunk` and `PutBlock` won't complete before the assertion. 
In `testWriteMoreThanFlushSize `, I think a flush is necessary to for the test. Also I add comments for the test.

Reply:
Thanks @chungen0126 for double-checking.  In that case, my results must have been coincidental.